src/osd/PG.cc

```diff
@@ -2622,131 +2629,150 @@ void PG::_update_calc_stats()
+    int64_t misplaced = 0;
+    // a degraded objects has fewer replicas or EC shards than the pool specifies.
+    int64_t degraded = 0;
+
+    for (auto m = missing_target_objects.rbegin();
+        m != missing_target_objects.rend(); ++m) {
 
-        osd_objects = MAX(0, num_objects - osd_missing);
-        object_copies += osd_objects;
-        // Count non-missing objects not in up as misplaced
-	if (!in_up) {
-	  misplaced += osd_objects;
-	  ++num_misplaced;
+      int64_t extra_missing = -1;
+
+      if (pool.info.is_replicated()) {
+	if (!acting_source_objects.empty()) {
+	  auto extra_copy = acting_source_objects.begin();
+	  extra_missing = std::get<0>(*extra_copy);
+          acting_source_objects.erase(*extra_copy);
+	}
+      } else {	// Erasure coded
+	// Use corresponding shard
+	for (const auto& a : acting_source_objects) {
+	  if (std::get<1>(a).shard == std::get<1>(*m).shard) {
+	    extra_missing = std::get<0>(a);
+	    acting_source_objects.erase(a);
+	    break;
+	  }
 	}
-      } else {
-        // If this peer has more objects then it should, ignore them
-        int64_t osd_backfilled = MIN(num_objects,
-				     peer_info[p].stats.stats.sum.num_objects);
-	backfill_target_objects.insert(make_pair(osd_backfilled, p));
-	backfilled += osd_backfilled;
       }
-    }
 
-    // Only include backfill targets below pool size into the object_copies
-    // count.  Use the most-full targets.
-    int num_backfill_shards_seen = 0;
-    for (auto i = backfill_target_objects.rbegin();
-	 i != backfill_target_objects.rend();
-	 ++i) {
-      if (actingset.size() + num_backfill_shards_seen < pool.info.size) {
-	object_copies += i->first;
-	++num_backfill_shards_seen;
+      if (extra_missing >= 0 && std::get<0>(*m) >= extra_missing) {
+	// We don't know which of the objects on the target
+	// are part of extra_missing so assume are all degraded.
+	misplaced += std::get<0>(*m) - extra_missing;
+	degraded += extra_missing;
       } else {
-	break;
+	// 1. extra_missing == -1, more targets than sources so degraded
+	// 2. extra_missing > std::get<0>(m), so that we know that some extra_missing
+	//    previously degraded are now present on the target.
+	degraded += std::get<0>(*m);
       }
     }
-
-    // Any objects that have been backfilled to up OSDs can deducted from misplaced
-    // adjusted by the assumption that backfill is happening approximately evenly
-    // across backfill nodes.  Use the most-full targets.
-    int64_t adjust_misplaced = 0;
-    for (auto i = backfill_target_objects.rbegin();
-	 i != backfill_target_objects.rend() && num_misplaced;
-	 ++i, --num_misplaced) {
-      adjust_misplaced += i->first;
+    // If there are still acting that haven't been accounted for
+    // then they are misplaced
+    for (const auto& a : acting_source_objects) {
+      dout(20) << __func__ << " extra acting misplaced " << std::get<0>(a) << dendl;
+      misplaced += std::max((int64_t)0, num_objects - std::get<0>(a));
```

Comment:
I'm not sure about this.  If you go from, say, 3x -> 2x replication, I don't think the 'misplaced' should show up at all--they're just stray objects that are queued for deletion.

I think we should define "misplaced" as a missing object (a copy an object not in its proper location) for which there is a corresponding copy/source of it elsewhere that is up to date.  For the 3x->2x (and this extra acting_source_objects entries in general) case, they are just extra copies floating around that aren't "used" as the source of a missing.

?

Reply:
Misplaced are all copies not in the proper place (acting only) for which there isn't a corresponding copy in the proper place (up).  But in a simple 3x -> 2x case there won't be misplaced because up/acting will immediately be reduced to 2 OSDs.  We don't care about when the unnecessary PG is deleted.

In the case of taking 1 OSD out, misplaced will be the total of all objects on the out OSD, but as they are backfilled to a new OSD the misplaced is reduced.  We don't care when the PG on the out OSD is deleted.

Does that sound reasonable?

One of the new tests hits the "extra acting misplaced" code path.  With 4 OSDs pool size 3, the test marks all 3 OSDs [1,2,0] out and sets the size of the pool to 2.  Even though I delay recovery to give time to see all the map changes, I still see this state with 3 acting OSDs for less than a second.  The total misplaced is 600 (200 objects on 3 acting OSDs and OSD 3 wanting to backfill.

pg[1.0( v 21'200 (21'100,21'200] local-lis/les=27/28 n=200 ec=18/18 lis/c 27/18 les/c/f 28/20/0 27/27/18) [3]/[1,2,0] r=0 lpr=27 pi=[18,27)/1 bft=3 crt=21'200 lcod 21'199 mlcod 0'0 active+remapped+backfilling] _update_calc_stats extra acting misplaced 200
pg[1.0( v 21'200 (21'100,21'200] local-lis/les=27/28 n=200 ec=18/18 lis/c 27/18 les/c/f 28/20/0 27/27/18) [3]/[1,2,0] r=0 lpr=27 pi=[18,27)/1 bft=3 crt=21'200 lcod 21'199 mlcod 0'0 active+remapped+backfilling] _update_calc_stats misplaced 600

From a practical stand point the user probably won't ever notice this very brief internal state.

The backfill sees the size change and restarts without any extra OSDs.

pg[1.0( v 21'200 (21'100,21'200] local-lis/les=30/31 n=200 ec=18/18 lis/c 30/18 les/c/f 31/20/0 27/30/18) [3]/[1,2] r=0 lpr=30 pi=[18,30)/1 rops=1 bft=3 crt=21'200 lcod 21'199 mlcod 0'0 active+remapped+backfilling] _update_calc_stats misplaced 400

We end with remapped state and 200 misplaced because osd 1 is marked out.

pg[1.0( v 21'200 (21'195,21'200] local-lis/les=32/33 n=200 ec=18/18 lis/c 32/32 les/c/f 33/33/0 27/32/32) [3]/[3,1] r=0 lpr=32 crt=21'200 mlcod 0'0 active+clean+remapped] _update_calc_stats misplaced 200

Reply:
Makes sense!